sklearn/linear_model/logistic.py

```diff
@@ -1557,32 +1554,34 @@ def fit(self, X, y, sample_weight=None):
 
         X, y = check_X_y(X, y, accept_sparse='csr', dtype=np.float64,
                          order="C")
-
+        check_classification_targets(y)
+        self.classes_ = np.unique(y)
         if self.solver == 'sag':
             max_squared_sum = row_norms(X, squared=True).max()
         else:
             max_squared_sum = None
 
-        check_classification_targets(y)
-
-        if y.ndim == 2 and y.shape[1] == 1:
-            warnings.warn(
-                "A column-vector y was passed when a 1d array was"
-                " expected. Please change the shape of y to "
-                "(n_samples, ), for example using ravel().",
-                DataConversionWarning)
-            y = np.ravel(y)
-
-        check_consistent_length(X, y)
+        if y.dtype.kind in ('S', 'U'):
+            # Encode for string labels
+            self._label_encoder = LabelEncoder().fit(y)
```

Comment:
Why can't you do encoding in all cases and assume that `LabelEncoder` handles efficiency issues?